File: Padrao Criacionais/FactoryMethod/FactoryMethod01.py
# Classe base de veículo
class Veiculo:
    def __init__(self, modelo):
        self.modelo = modelo

    def mostrar_detalhes(self):
        print(f"Modelo: {self.modelo}")

# Subclasses de veículos
class Carro(Veiculo):
    def __init__(self, modelo):
        super().__init__(modelo)

class Moto(Veiculo):
    def __init__(self, modelo):
        super().__init__(modelo)

class Barco(Veiculo):
    def __init__(self, modelo):
        super().__init__(modelo)

# Fábrica Abstrata de veículos
from abc import ABC, abstractmethod

class FabricaDeVeiculo(ABC):
    @abstractmethod
    def criar_veiculo(self, modelo):
        pass

# Fábrica Concreta de Carros
class FabricaDeCarros(FabricaDeVeiculo):
    def criar_veiculo(self, modelo):
        return Carro(modelo)

# Fábrica Concreta de Motos
class FabricaDeMotos(FabricaDeVeiculo):
    def criar_veiculo(self, modelo):
        return Moto(modelo)

# Fábrica Concreta de Barcos
class FabricaDeBarcos(FabricaDeVeiculo):
    def criar_veiculo(self, modelo):
        return Barco(modelo)

# Uso do Factory Method
fabrica_de_carros = FabricaDeCarros()
carro1 = fabrica_de_carros.criar_veiculo('SUV')
carro2 = fabrica_de_carros.criar_veiculo('4x4')
carro1.mostrar_detalhes()
carro2.mostrar_detalhes()

fabrica_de_motos = FabricaDeMotos()
moto = fabrica_de_motos.criar_veiculo('Esportivo')
moto.mostrar_detalhes()

fabrica_de_barcos = FabricaDeBarcos()
barco = fabrica_de_barcos.criar_veiculo('Náutico')
barco.mostrar_detalhes()

File: Padrao Criacionais/FactoryMethod/FactoryMethod01.java
public class FactoryMethod01 {

    // Classe base de veículo
    public abstract static class Veiculo {
        private String modelo;

        public Veiculo(String modelo) {
            this.modelo = modelo;
        }

        public void mostrarDetalhes() {
            System.out.println("Modelo: " + modelo);
        }
    }

    // Subclasses de veículos
    public static class Carro extends Veiculo {
        public Carro(String modelo) {
            super(modelo);
        }
    }

    public static class Moto extends Veiculo {
        public Moto(String modelo) {
            super(modelo);
        }
    }

    public static class Barco extends Veiculo {
        public Barco(String modelo) {
            super(modelo);
        }
    }

    // Fábrica Abstrata de veículos
    public abstract static class FabricaDeVeiculo {
        public abstract Veiculo criarVeiculo(String modelo);
    }

    // Fábrica Concreta de Carros
    public static class FabricaDeCarros extends FabricaDeVeiculo {
        public Veiculo criarVeiculo(String modelo) {
            return new Carro(modelo);
        }
    }

    // Fábrica Concreta de Motos
    public static class FabricaDeMotos extends FabricaDeVeiculo {
        public Veiculo criarVeiculo(String modelo) {
            return new Moto(modelo);
        }
    }

    // Fábrica Concreta de Barcos
    public static class FabricaDeBarcos extends FabricaDeVeiculo {
        public Veiculo criarVeiculo(String modelo) {
            return new Barco(modelo);
        }
    }

    // Uso do Factory Method
    public static void main(String[] args) {
        FabricaDeVeiculo fabricaDeCarros = new FabricaDeCarros();
        Veiculo carro1 = fabricaDeCarros.criarVeiculo("SUV");
        Veiculo carro2 = fabricaDeCarros.criarVeiculo("4x4");
        carro1.mostrarDetalhes();
        carro2.mostrarDetalhes();

        FabricaDeVeiculo fabricaDeMotos = new FabricaDeMotos();
        Veiculo moto = fabricaDeMotos.criarVeiculo("Esportivo");
        moto.mostrarDetalhes();

        FabricaDeVeiculo fabricaDeBarcos = new FabricaDeBarcos();
        Veiculo barco = fabricaDeBarcos.criarVeiculo("Náutico");
        barco.mostrarDetalhes();
    }
}